src/Workspaces/CSharp/Portable/Classification/CSharpClassificationProvider.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Classification.Classifiers;
+using Microsoft.CodeAnalysis.CSharp.Classification.Classifiers;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Classification
+{
+    [ExportClassificationProvider(ClassificationProviderNames.Default, LanguageNames.CSharp), Shared]
+    internal class CSharpClassificationProvider : CommonClassificationProvider
+    {
+        private static readonly ImmutableArray<ISemanticClassifier> s_defaultClassifiers =
+            ImmutableArray.Create<ISemanticClassifier>(
+                new NameSyntaxClassifier(),
+                new SyntaxTokenClassifier(),
+                new UsingDirectiveSyntaxClassifier());
+
+        public override ImmutableArray<ISemanticClassifier> GetDefaultSemanticClassifiers()
+        {
+            return s_defaultClassifiers;
+        }
+
+        public override void AddLexicalClassifications(SourceText text, TextSpan span, ClassificationContext context, CancellationToken cancellationToken)
+        {
+            ClassificationHelpers.AddLexicalClassifications(text, span, context, cancellationToken);
+        }
+
+        protected override void AddSyntacticClassifications(SyntaxTree syntaxTree, TextSpan textSpan, ClassificationContext context, CancellationToken cancellationToken)
+        {
+            var root = syntaxTree.GetRoot(cancellationToken);
+            SyntacticClassifier.CollectClassifiedSpans(root, textSpan, context, cancellationToken);
+        }
+
+        public override ClassifiedSpan AdjustClassification(SourceText rawText, ClassifiedSpan classifiedSpan)
```

Comment:
Out of curiosity, does this need to be on the top level service (or public)?  @paulvanbrenk you don't ever implement this right?

I think this is more for C# and VB to have their really crazy impl that attempts to reuse old results and then patch them up.  I think we could make this internal.